keras_nlp/utils/pipeline_model.py

```diff
@@ -0,0 +1,149 @@
+# Copyright 2022 The KerasNLP Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""A base class for models including preprocessing."""
+
+import tensorflow as tf
+from tensorflow import keras
+
+
+def convert_inputs_to_dataset(
+    x=None,
+    y=None,
+    sample_weight=None,
+    batch_size=None,
+):
+    if isinstance(x, tf.data.Dataset):
+        if y is not None:
+            raise ValueError(
+                "When `x` is a `tf.data.Dataset`, please do not provide "
+                f"`y`. Received: `y={y}`."
+            )
+        if sample_weight is not None:
+            raise ValueError(
+                "When `x` is a `tf.data.Dataset`, please do not provide "
+                f"`sample_weight`. Received: `sample_weight={sample_weight}`."
+            )
+        return x
+
+    inputs = keras.utils.pack_x_y_sample_weight(x, y, sample_weight)
+    return tf.data.Dataset.from_tensor_slices(inputs).batch(batch_size or 32)
+
+
+def apply_dataset_fn(ds, fn):
+    return ds.map(fn, tf.data.AUTOTUNE).prefetch(tf.data.AUTOTUNE)
+
+
+class PipelineModel(keras.Model):
+    """A model which allows automatically applying preprocessing."""
+
+    def __init__(self, *args, include_preprocessing=True, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.include_preprocessing = include_preprocessing
+
+    def preprocess_features(self, x):
```

Comment:
In a fully generalizable API we are likely going to need a `training` arg here and below

Reply:
I think this might be easier to do when upstreaming into core Keras. Getting `training` correct at call time seems to rely on some internal stuff, and we don't really need it on the NLP side so much I think.

https://github.com/keras-team/keras/blob/master/keras/engine/base_layer.py#L1056